Sources/OpenAI/Public/Models/Chat/ChatTool.swift

```diff
@@ -0,0 +1,191 @@
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        var dynamicContainer = encoder.container(keyedBy: DynamicKey.self)
+
+        try container.encode(type, forKey: .type)
+        
+        switch value {
+        case .function(let function):
+            try dynamicContainer.encode(function, forKey: .init(stringValue: "function"))
+            break
+        }
+    }
+    
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let dynamicContainer = try decoder.container(keyedBy: DynamicKey.self)
+        self.type = try container.decode(ChatTool.ToolType.self, forKey: .type)
+        
+        switch self.type {
+        case .function:
+            self.value = try dynamicContainer.decode(ChatTool.ToolValue.self, forKey: .init(stringValue: "function"))
+            break
+        }
+    }
+}
+
+
+extension ChatTool {
+    public enum ToolType: String, Codable {
+        case function
+    }
+    
+    public struct Function: Codable, Equatable {
+        /// The name of the function to be called. Must be a-z, A-Z, 0-9, or contain underscores and dashes, with a maximum length of 64.
+        public let name: String
+        
+        /// The parameters the functions accepts, described as a JSON Schema object. See the guide for examples, and the JSON Schema reference for documentation about the format.
+        public let parameters: JSONSchema
+        
+        /// A description of what the function does, used by the model to choose when and how to call the function.
+        public let description: String?
+        
+        public init(name: String, description: String? = nil, parameters: JSONSchema = .empty) {
+            self.name = name
+            self.parameters = parameters
+            self.description = description
+        }
+    }
+    
+    public enum ToolValue: Codable, Equatable {
+        case function(Function)
+        
+        public func encode(to encoder: Encoder) throws {
+            var container = encoder.singleValueContainer()
+            
+            switch self {
+            case .function(let function):
+                try container.encode(function)
+                break
+            }
+        }
+        
+        
+        public init(from decoder: Decoder) throws {
+            let container = try decoder.singleValueContainer()
+            
+            if let function = try? container.decode(Function.self) {
+                self = .function(function)
+            }  else {
+                throw DecodingError.dataCorrupted(
+                    DecodingError.Context(
+                        codingPath: decoder.codingPath,
+                        debugDescription: "Invalid data encountered when decoding StringOrCodable<T>"
+                    )
+                )
+            }
+        }
+    }
+}
+
+public struct ToolCall: Codable, Equatable {
+    public let index: Int
```

Comment:
I'm getting decoding errors because `index` is not present in the response.
```
▿ DecodingError
  ▿ keyNotFound : 2 elements
    - .0 : CodingKeys(stringValue: "index", intValue: nil)
    ▿ .1 : Context
      ▿ codingPath : 5 elements
        - 0 : CodingKeys(stringValue: "choices", intValue: nil)
        ▿ 1 : _JSONKey(stringValue: "Index 0", intValue: 0)
          ▿ rep : Rep
            - index : 0
        - 2 : CodingKeys(stringValue: "message", intValue: nil)
        - 3 : CodingKeys(stringValue: "tool_calls", intValue: nil)
        ▿ 4 : _JSONKey(stringValue: "Index 0", intValue: 0)
          ▿ rep : Rep
            - index : 0
      - debugDescription : "No value associated with key CodingKeys(stringValue: \"index\", intValue: nil) (\"index\")."
      - underlyingError : nil
```